- Raw coverage data ("TCGA_WES_hg19_N.rds" and "TCGA_WES_hg19_T.rds") were generated by running GATK DepthOfCoverage tool on TCGA WES bam files (GRCh37) stored in Firecloud/Terra and output files were downloaded for downstream analyses with R.

Terra workspace
https://app.terra.bio/#workspaces/gdan-copy-number-gdac/Tangent_TCGA_SexChromosomes_KE_2

WDL
02_TCGA_data_preparation/src/DepthOfCoverage_TCGA.wdl

- "whole_exome_agilent_1.1_refseq_plus_3_boosters.targetIntervals.bed" were downloaded from the following link
https://bitbucket.org/cghub/cghub-capture-kit-info/src/b33c1cb33593/BI/vendor/Agilent/whole_exome_agilent_1.1_refseq_plus_3_boosters.targetIntervals.bed?at=master

- "PAR.txt" was generated based on information obtained from wikipedia page on pseudoautosomal region.
https://en.wikipedia.org/wiki/Pseudoautosomal_region

- "mmc2.xlsx" was obtained from supplementary data of Sadagopan et al., 2022 Cell Systems "Somatic XIST activation and features of X chromosome inactivation in male human cancers".

- "1-s2.0-S0092867423006463-mmc1.xlsx" was obtained from supplemantary data of Qi et al., 2023 Cell "Loss of chromosome Y in primary tumors".

- "CNV.hg19.bypos.111213.txt" was obtained from Tangent github page.
https://github.com/broadinstitute/tangent/tree/master/lib

- RNA seq expression data were downloaded from GDC Data Portal using GDC Data Transfer Tool with manifest file "gdc_manifest_TCGA_RNAseq.txt".
